docs/developer-docs/security/security-best-practices/data-integrity-and-authenticity.mdx

```diff
@@ -8,19 +8,334 @@ import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";
+
+:::tip
+ICP also provides Replica signed queries, where query responses are signed by the answering replica node but however it doesn't have the same security guarantees as `update` calls. Replica signed queries are enabled by default on both `agent-rs` and `agent-js`. You can read more about the feature [here](https://internetcomputer.org/blog/features/replica-signed-queries)
+:::
+
+### What's Certified data?
+* A canister can store a small amount of data (32 bytes) during an update call in the state tree under `certified_data` which will be certified by the subnet.
+* During query call, the canister can obtain a certificate about that data, which can be verified by the IC root public key.
+
+In the rest of the section, we will see an example of how to design a canister, which can serve a certified response for a `query` using `certified_data` and how to verify the response. The examples are written in Rust but can be imported to other languages following the same recommendations.
+
+### Building a canister with Certified Variables
+Let's consider the following canister,
+
+```c
+type User = record {
+    name: text;
+    age: nat8;
+};
+
+type CertifiedUser = record {
+    user : User;
+    certificate : blob;
+    witness : blob;
+};
+
+service : {
+    "set_user": (User) -> (nat64);
+    "get_user": (nat64) -> (CertifiedUser) query;
+}
+```
+
+The canister exposes the following service, 
+* **set_user** - The caller provides a `User` object to the canister. The canister records it and serves a corresponding `index` for the entry as the response. Since `certified_data` can only store 32 bytes of data, we use a specialized data strucutre from `ic_certified_map` to store the `User` data. 
+  * The data strucutre internally stores the data in a `HashTree` and we record the `root_hash` of the data structure in the `certified_data`, which is 32 bytes. 
+  * The `root_hash` cryptographically guarantees that only one tree can correspond to that hash. You can read more about it [here](https://docs.rs/ic-certified-map/latest/ic_certified_map/struct.RbTree.html).
```

Comment:
The referenced documentation doesn't really explain HashTrees. Is there a better source for people to understand how a 32 bytes hash can certify an endless datastructure?

Reply:
I don't think we have anything explaining Merkle Trees.  I can reword it a bit and point to Merkle trees Wiki page.